galleries/examples/user_interfaces/embedding_in_qt_sgskip.py

```diff
@@ -44,18 +44,34 @@ def __init__(self):
         self._static_ax.plot(t, np.tan(t), ".")
 
         self._dynamic_ax = dynamic_canvas.figure.subplots()
-        t = np.linspace(0, 10, 101)
         # Set up a Line2D.
-        self._line, = self._dynamic_ax.plot(t, np.sin(t + time.time()))
-        self._timer = dynamic_canvas.new_timer(50)
-        self._timer.add_callback(self._update_canvas)
-        self._timer.start()
+        self._update_data()
+        self._line, = self._dynamic_ax.plot(self.xdata, self.ydata)
+        # The below two timers must be attributes of self, so that the garbage
+        # collector won't clean them after we finish with __init__...
 
-    def _update_canvas(self):
-        t = np.linspace(0, 10, 101)
+        # The data retrieval should be fast as possible (Using QRunnable should
+        # be even faster)
+        self.dataTimer = dynamic_canvas.new_timer(1)
+        self.dataTimer.add_callback(self._update_data)
+        self.dataTimer.start()
+        # Drawing at 50Hz should be fast enough for the GUI to feel smooth, and
+        # not too fast for the GUI to be overloaded with events that need to be
+        # processed while the GUI element is changed.
+        self.drawingTimer = dynamic_canvas.new_timer(50)
+        self.drawingTimer.add_callback(self._update_canvas)
+        self.drawingTimer.start()
+
+    def _update_data(self):
+        self.xdata = np.linspace(0, 10, 101)
```

Comment:
To check my understanding, because these are being run by timmers, that means they are run atomically on the event loop so we do not need to worry about locking between this method and `_update_canvas`?

Reply:
> To check my understanding, because these are being run by timmers, that means they are run atomically on the event loop so we do not need to worry about locking between this method and `_update_canvas`?

I'm not sure what do you mean by locking, are you talking about potential issues of xdata not in sync with ydata or such?

Reply:
I'm worried about the sequence of things where:

 - you set `self.xdata`
 - `self._line.set_data(self.xdata, self.ydata)` runs
 - you set `self.ydata`

In this particular example, if we were losing that race it would be OK because the xdata is always the same (important from a "correct plot" point of view) and the same length (important from a "the code will run without raising" point of view).   If there are threads involved the above sequence is possible as with non-free threading Python (aka all released versions) the interpreter can drop the GIL after any bytecode evaluation and then any thread can be the next one to run (and in the future with free-threading builds of Python it is just the YOLO land).

However, if the calls to these methods are serialized by the event loop (it runs the one at a time in the order they were put on the queue) then we are safe because Qt is providing the level of locking we need.

Reply:
OK I see what you mean. Currently the `dataTimer` starts (with `.start()`) before the `drawingTimer` starts, and also the timeout of the 1st timer for sure is reached before the timeout of the 2nd timer (due to the different intervals), so I am pretty sure we can count on that.

Not only that, even if both timers had the same interval, the `drawingTimer`'s callback that uses `self._line.set_data(self.xdata, self.ydata)` always has a `self.{x,y}data` attributes available, because I ran manually the `self._update_data` method before I create the line with `self._dynamic_ax.plot`.

Reply:
Consider the case where : 

```
import random
self.xdata = np.linspace(0, 10, random.randint(100, 150))
```

Which callback run first or if the attributes exist is not the issue, the chance of them running concurrently is the problem!

Reply:
> Consider the case where :
> 
> ```
> import random
> self.xdata = np.linspace(0, 10, random.randint(100, 150))
> ```
> 
> Which callback run first or if the attributes exist is not the issue, the chance of them running concurrently is the problem!

Hmm I'm pretty sure that the timers' callbacks run concurrently, and they block the event loop - that's my main conclusion of the investigation I performed on the "stuck" issue I experienced. In the case you presented, at the same callback that sets `self.xdata`, `self.ydata` should be set as well, to make sure they have the same size, and then when the blocking callback of the drawing happens - they will still have the same size.

Perhaps it is worth explaining in a comment? Or perhaps I will set `self.xdata` in the `_update_data` callback to the same value, although it is not really needed, just for the sake of the example?